### PR TITLE
Add ffmpeg-sidecar support

### LINKED ISSUE #194: switch to ffmpeg-sidecar
https://github.com/CapSoftware/ffmpeg-sidecar

https://github.com/CapSoftware/Cap/blob/63719073805f1fd040343ae27366272a29918078/apps/desktop/src-tauri/src/main.rs#L75

switch to use this lib to install ffmpeg in screenpipe lib

Update: little concern with cap fork is that it's using their s3 ffmpeg build so if they close it, product is down, but can work around this later i guess

### PR DESCRIPTION
**Pull request by Neptune650**
/claim #194
/closes #194

This PR adds ffmpeg-sidecar with auto install, it is partly inspired by the older PR but I've tested it and made significant important changes. Fork has been changed due to being more updated.

### CODE REVIEW

**Comment by Neptune650** [comment]
@louis030195 Please check this https://github.com/mediar-ai/screenpipe/issues/194#issuecomment-2439041973 and review.

**Comment by Neptune650** [comment]
Hello, I've done the tests you requested

With h265:
image

With h264:
image

**Comment by Neptune650** [comment]
@louis030195 
As you can see, when we use H265 for FFmpeg encoding - while resource usage only increases by around 1%, it brings huge benefits to storage usage, with a decrease of a **11%**, this can matter a lot for extremely long duration files.

And another benefit is that we no longer need to use a different encoding for Windows and other platforms - we can just use libx265 for every platform.

File: screenpipe-server/src/bin/screenpipe-server.rs [hunk 1]
```diff
@@ -174,7 +174,12 @@ async fn main() -> anyhow::Result<()> {
                 }
 
                 info!("screenpipe setup complete");
-                // TODO: ffmpeg sidecar thing here
+                if let Some(ffmpeg_path) = find_ffmpeg_path() {
```
Comment (louis030195) [inline screenpipe-server/src/bin/screenpipe-server.rs:177]: we should download ffmpeg here using the lib, as this is used in app onboarding

File: screenpipe-server/src/video.rs [hunk 2]
```diff
@@ -169,21 +169,7 @@ async fn start_ffmpeg_process(output_file: &str, fps: f64) -> Result<Child, anyh
...
-            "-q:v",
-            "5", // Adjust quality (1-31, lower is better)
-            "-preset",
-            "ultrafast",
-        ]);
-    } else {
-        // Use libx264 for other platforms
-        args.extend_from_slice(&["-vcodec", "libx264", "-preset", "ultrafast", "-crf", "23"]);
-    }
+    args.extend_from_slice(&["-vcodec", "libx265", "-preset", "ultrafast", "-crf", "23"]);
```
Comment (louis030195) [inline screenpipe-server/src/video.rs:172]: can you confirm we can still render video on windows in the app chat?

i remember chatgpt telling me using h265 risks some device not being able to render video

Comment (louis030195) [reply screenpipe-server/src/video.rs:172]: e.g. just press button search and see if videos render

File: screenpipe-server/src/bin/screenpipe-server.rs (hunk 1 above)
Comment (Neptune650) [reply screenpipe-server/src/bin/screenpipe-server.rs:177]: @louis030195 Look, it already does that: https://github.com/mediar-ai/screenpipe/blob/704762f52279fd4fbc1e7a2d52835ded18e96f46/screenpipe-core/src/ffmpeg.rs#L87

File: screenpipe-server/src/video.rs (hunk 2 above)
Comment (Neptune650) [reply screenpipe-server/src/video.rs:172]: @louis030195 I tested and it works... H265 support is pretty widespread now. AFAIK every device with Windows 10 should be able to run it.

**Comment by louis030195** [comment]
still think we can remove old ffmepg then

**Comment by Neptune650** [comment]
> still think we can remove old ffmepg then

@louis030195 I think it's unnecessary - it just stays as a fallback and doesn't cause any issues at all.

**Comment by louis030195** [comment]
```
(env) (base) louisbeaumont@mac:~/Documents/screen-pipe$ cargo build --release --features metal
    Updating git repository `https://github.com/nathanbabcock/ffmpeg-sidecar`
    Updating crates.io index
error: failed to select a version for `tar`.
    ... required by package `ort-sys v2.0.0-rc.6`
    ... which satisfies dependency `ort-sys = "^2.0.0-rc.5"` (locked to 2.0.0-rc.6) of package `ort v2.0.0-rc.5`
    ... which satisfies dependency `ort = "^2.0.0-rc.5"` (locked to 2.0.0-rc.5) of package `screenpipe-audio v0.1.97 (/Users/louisbeaumont/Documents/screen-pipe/screenpipe-audio)`
    ... which satisfies path dependency `screenpipe-audio` (locked to 0.1.97) of package `screenpipe-server v0.1.97 (/Users/louisbeaumont/Documents/screen-pipe/screenpipe-server)`
versions that meet the requirements `^0.4` (locked to 0.4.41) are: 0.4.41

all possible versions conflict with previously selected packages.

  previously selected package `tar v0.4.42`
    ... which satisfies dependency `tar = "^0.4.42"` of package `ffmpeg-sidecar v2.0.0-rc (https://github.com/nathanbabcock/ffmpeg-sidecar?branch=main#3fd9caff)`
    ... which satisfies git dependency `ffmpeg-sidecar` of package `screenpipe-core v0.1.97 (/Users/louisbeaumont/Documents/screen-pipe/screenpipe-core)`
    ... which satisfies path dependency `screenpipe-core` (locked to 0.1.97) of package `screenpipe-server v0.1.97 (/Users/louisbeaumont/Documents/screen-pipe/screenpipe-server)`

```

**Comment by louis030195** [comment]
```
(env) (base) louisbeaumont@mac:~/Documents/screen-pipe$ ./target/release/screenpipe setup
2024-10-30T00:55:25.278789Z  INFO screenpipe: logging initialized
2024-10-30T00:55:25.455469Z  INFO screenpipe: Audio permission requested. Please grant permission if prompted.
2024-10-30T00:55:55.535208Z  INFO screenpipe: Screen capture permission requested. Please grant permission if prompted.
2024-10-30T00:55:55.541221Z  INFO screenpipe_audio::whisper: device = Metal(MetalDevice(DeviceId(1)))    
2024-10-30T00:55:57.967964Z  INFO screenpipe_audio::vad_engine: Initializing SileroVad...
2024-10-30T00:55:57.968470Z  INFO screenpipe_audio::vad_engine: SileroVad Model downloaded to: "/Users/louisbeaumont/Library/Caches/screenpipe/vad/silero_vad.onnx"
2024-10-30T00:55:58.024703Z ERROR screenpipe: FFmpeg check failed: No such file or directory (os error 2)
2024-10-30T00:55:58.024999Z ERROR screenpipe: Please ensure FFmpeg is installed correctly and is in your PATH
Error: No such file or directory (os error 2)
```

**Comment by louis030195** [comment]
https://github.com/Neptune650/screenpipe/blob/704762f52279fd4fbc1e7a2d52835ded18e96f46/screenpipe-server/src/bin/screenpipe-server.rs#L167


did you run the code yourself?

**Comment by Neptune650** [comment]
> https://github.com/Neptune650/screenpipe/blob/704762f52279fd4fbc1e7a2d52835ded18e96f46/screenpipe-server/src/bin/screenpipe-server.rs#L167
> 
> did you run the code yourself?

@louis030195 Oh - my bad, I hadn't checked with screenpipe setup.
Pushing a fix.

**Comment by Neptune650** [comment]
@louis030195 It doesn't cause any issues now.

**Comment by Neptune650** [comment]
> `tar = "=0.4.42"`

I hadn't seen this error in Linux or Windows, adding it.

**Comment by louis030195** [comment]
image

image


🥹


that was my concern

**Comment by Neptune650** [comment]
> so i assume we need to install some video rendering lib for h265 in tauri webview
@louis030195 Any suggestion to how we could do that

**Comment by louis030195** [comment]
@Neptune650 

https://github.com/tauri-apps/tauri/issues/11559

**Comment by louis030195** [comment]
chatgpt says

```tsx
```screenpipe-app-tauri/components/video.tsx
// ... existing code ...
const getMimeType = (path: string): string => {
    const ext = path.split(".").pop()?.toLowerCase();
    switch (ext) {
        case "mp4":
            return "video/mp4";
        case "webm":
            return "video/webm";
        case "ogg":
            return "video/ogg";
        case "mp3":
            return "audio/mpeg";
        case "wav":
            return "audio/wav";
        case "h265": // Add support for H.265
            return "video/h265"; // Adjust MIME type if necessary
        default:
            return isAudio ? "audio/mpeg" : "video/mp4";
    }
};
// ... existing code ...
```

will try when i have time

**Comment by Neptune650** [comment]
> chatgpt says
> 
> ```tsx
> ```screenpipe-app-tauri/components/video.tsx
> // ... existing code ...
> const getMimeType = (path: string): string => {
>     const ext = path.split(".").pop()?.toLowerCase();
>     switch (ext) {
>         case "mp4":
>             return "video/mp4";
>         case "webm":
>             return "video/webm";
>         case "ogg":
>             return "video/ogg";
>         case "mp3":
>             return "audio/mpeg";
>         case "wav":
>             return "audio/wav";
>         case "h265": // Add support for H.265
>             return "video/h265"; // Adjust MIME type if necessary
>         default:
>             return isAudio ? "audio/mpeg" : "video/mp4";
>     }
> };
> // ... existing code ...
> ```
> 
> (seems wrong but need to investigate mime type)
> 
> will try when i have time

@louis030195 I'll also carry out my own tests

**Comment by louis030195** [comment]
tried again this PR to fix macos that does not render:
- using `react-player`
- using `plyr-react`
- tweaking the mime type

does not work still

**Comment by Neptune650** [comment]
> tried again this PR to fix macos that does not render:
> 
>     * using `react-player`
> 
>     * using `plyr-react`
> 
>     * tweaking the mime type
> 
> 
> does not work still

@louis030195 How do you feel about keeping the ffmpeg sidecar but using x264?

**Comment by louis030195** [comment]
the point of this was to enable h265

users don't have ffmpeg issues with the current build

**Comment by Neptune650** [comment]
> the point of this was to enable h265
> 
> users don't have ffmpeg issues with the current build

@louis030195 Well other benefits is that it auto-installs ffmpeg if not present, and also can use x264 in windows instead of x264_mf

**Comment by louis030195** [comment]
@Neptune650 okay sure, let's do this?

lib265 for linux and windows
lib264 for macos (until we figure how to fix the rendering in tauri)

**Comment by Neptune650** [comment]
> @Neptune650 okay sure, let's do this?
> 
> lib265 for linux and windows lib264 for macos (until we figure how to fix the rendering in tauri)

@louis030195 Sounds good to me, thanks.

**Pull request merged**